northstar/src/runtime/minijail.rs

```diff
@@ -509,84 +521,93 @@ impl Debug {
+/// Creates a pipe with the read endpoint implementing tokio::io::AsyncRead
+async fn create_pipe() -> Result<(PipeRead, PipeWrite), Error> {
+    let (readfd, writefd) =
+        pipe2(OFlag::O_NONBLOCK).map_err(|e| Error::Os("Failed to create pipe".to_string(), e))?;
+    let async_read =
+        AsyncFd::new(PipeFd(readfd)).map_err(|e| Error::Io("Async fd".to_string(), e))?;
+    Ok((PipeRead(async_read), PipeWrite(PipeFd(writefd))))
 }
 
-impl AsyncPipe {
-    fn new() -> Result<AsyncPipe, Error> {
-        let (non_blocking, blocking) =
-            pipe().map_err(|e| Error::Os("Failed to create pipe".to_string(), e))?;
-
-        let mut flags =
-            OFlag::from_bits(fcntl(non_blocking, fcntl::FcntlArg::F_GETFL).unwrap()).unwrap();
-        flags.set(OFlag::O_NONBLOCK, true);
-        fcntl(non_blocking, fcntl::FcntlArg::F_SETFL(flags)).expect("Failed to configure pipe fd");
-        let non_blocking = unsafe { std::fs::File::from_raw_fd(non_blocking) };
-        let non_blocking =
-            AsyncFd::new(non_blocking).map_err(|e| Error::Io("Async fd".to_string(), e))?;
-
-        Ok(AsyncPipe {
-            non_blocking,
-            blocking,
-        })
-    }
+#[derive(Debug)]
+struct PipeFd(RawFd);
 
-    fn blocking_fd(&self) -> RawFd {
-        self.blocking
+impl AsRawFd for PipeFd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.0
     }
 }
 
-impl Drop for AsyncPipe {
+impl Drop for PipeFd {
     fn drop(&mut self) {
-        unistd::close(self.blocking).ok();
+        if let Err(e) = unistd::close(self.0) {
+            error!("Failed to close pipe: {}", e);
+        }
     }
 }
 
-impl AsyncRead for AsyncPipe {
+/// Pipe with async tokio impls
+#[derive(Debug)]
+struct PipeRead(AsyncFd<PipeFd>);
+
+impl AsyncRead for PipeRead {
     fn poll_read(
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
         buf: &mut ReadBuf<'_>,
     ) -> Poll<io::Result<()>> {
         loop {
-            let mut guard = futures::ready!(self.non_blocking.poll_read_ready(cx))?;
-            match guard
-                .try_io(|inner| std::io::Read::read(&mut inner.get_ref(), buf.initialized_mut()))
-            {
+            let mut guard = futures::ready!(self.0.poll_read_ready(cx))?;
+            match guard.try_io(|inner| {
+                let fd = inner.get_ref().as_raw_fd();
+                // map nix::Error to io::Error
+                match unistd::read(fd, buf.initialized_mut()) {
+                    Ok(n) => Ok(n),
+                    // read(2) on a nonblocking file (O_NONBLOCK) returns EAGAIN or EWOULDBLOCK in
+                    // case that the read would block. That case is handled by `try_io`.
+                    Err(nix::Error::Sys(e)) => Err(std::io::Error::from_raw_os_error(e as i32)),
+                    Err(e) => Err(std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        format!("{}", e),
+                    )),
+                }
+            }) {
```

Comment:
Whats the advantage compared to the current version?

Reply:
I think this version is a bit more transparent when handling the pipe fd. Mainly trying to avoid any surprise when delegating this to `std::fs::File`'s implementation.